org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/parser/Parser.java

```diff
@@ -4874,6 +4876,55 @@ protected void consumeInternalCompilationUnit() {
 		this.compilationUnit.createPackageInfoType();
 	}
 }
+protected void consumeUnnamedClassBodyDeclarations() {
+	concatNodeLists();
+}
+
+protected void consumeInternalCompilationUnitWithPotentialUnnamedClass() {
+	int length;
+	if ((length = this.astLengthStack[this.astLengthPtr--]) != 0) {
+		LinkedList<AbstractMethodDeclaration> methods = new LinkedList<>();
+		LinkedList<FieldDeclaration> fields = new LinkedList<>();
+		LinkedList<TypeDeclaration> types = new LinkedList<>();
+		TypeDeclaration topLevelType = null;
+		// here are length declarations
+		for (int i = length - 1; i >= 0; i--) {
+			ASTNode astNode = this.astStack[this.astPtr--];
+			if (astNode instanceof MethodDeclaration method) {
+				//methods and constructors have been regrouped into one single list
+				methods.addFirst(method);
+			} else if (astNode instanceof TypeDeclaration type) {
+				if (!Arrays.equals(this.compilationUnit.getMainTypeName(), type.name)) {
+					types.addFirst(type);
+				} else {
+					topLevelType = type;
+				}
+			} else if (astNode instanceof FieldDeclaration field) {
+				fields.addFirst(field);
+			}
+		}
+		if ((!methods.isEmpty() || !fields.isEmpty()) && problemReporter().validateJavaFeatureSupport(JavaFeature.UNNAMMED_CLASSES_AND_INSTANCE_MAIN_METHODS, 0, 0)) {
```

Comment:
I think this should be `!problemReporter().validateJavaFeatureSupport(...`, since it seems like `validateJavaFeatureSupport` returns `true` when a diagnostic is reported and `false` otherwise.

Reply:
>mpalat requested a review from [srikanth-sankaran](https://github.com/srikanth-sankaran) [now](https://github.com/eclipse-jdt/eclipse.jdt.core/pull/1521#event-10939188376)

@srikanth-sankaran - adding you as a reviewer once you are done with String templates. thanks